proposals/NNNN-nonisolated-for-global-actor-cutoff.md

```diff
@@ -0,0 +1,337 @@
+
+#### Extensions
+
+This proposal allows for `nonisolated` attribute to be applied on extension declarations:
+
+```swift
+nonisolated extension GloballyIsolated {
+  var x: NonSendable { .init() }
+  func implicitlyNonisolated() {}
+}
+
+struct C: GloballyIsolated {
+  nonisolated func explicitlyNonisolated() {
+    let _ = x // okay
+    implicitlyNonisolated() // okay
+  }
+}
+```
+
+In the code above, the  `nonisolated` attribute is applied to an extension declaration for a `GloballyIsolated` protocol. When applied to an extension, `nonisolated` applies to all of its members. In this case, `implicitlyNonisolated` method and the computed property `x` are both nonisolated, and therefore are able to be accessed from a nonisolated context in the body of `explicitlyNonisolated` method of a globally-isolated struct `C`.
+
+#### Classes, structs, and enums
+
+Finally, we propose allowing writing `nonisolated` on class, struct and enum declarations:
+
+```swift
+nonisolated class K: GloballyIsolated {
+  var x: NonSendable
+  init(x: NonSendable) {
+    self.x = x // okay, 'x' is non-isolated
+  }
+} 
+
+nonisolated struct S: GloballyIsolated {
+  var x: NonSendable
+  init(x: NonSendable) {
+    self.x = x // okay, 'x' is non-isolated
+  }
+} 
+
+nonisolated enum E: GloballyIsolated {
+  func implicitlyNonisolated() {}
+  init() {}
+}
+
+struct TestEnum {
+  nonisolated func call() {
+    E().implicitlyNonisolated() // okay
+  }
+}
+```
+
+In all the above declarations, the `nonisolated` attribute propagates to all of their members, therefore making them accessible from a non-isolated context.
+
+Importantly, types nested inside of explicitly `nonisolated` declarations still infer actor isolation from their own conformance lists:
+
+```swift
+nonisolated struct S: GloballyIsolated {
+  var value: NotSendable // 'value' is not isolated
+  struct Nested: GloballyIsolated {} // 'Nested' is still @MainActor-isolated
+}
+```
+
+The above behavior is semantically consistent with the existing rules around global isolation inference for members of a type:
+
+```swift
+@MainActor struct S {
+  var value: NotSendable // globally-isolated
+  struct Nested {} // 'Nested' is not @MainActor-isolated
+}
+```
+
+### Annotating more types of storage with `nonisolated`
+
+This section extends the existing rules for when `nonisolated` can be written on a storage of a user-defined type.
+
+#### Stored properties of non-`Sendable` types
+
+Currently, any stored property of a non-`Sendable` type is implicitly treated as non-isolated. This proposal allows for spelling of this behavior:
+
+```swift
+class MyClass {
+  nonisolated var x: NonSendable = NonSendable() // okay
```

Comment:
Yes, I think the main goal here is to allow expressing the default behavior for consistency. I don't think I have a motivating use case here, since it's more about a personal preference of the programmer.